support/mesh/mesh_channel_core/src/deque.rs

```diff
@@ -0,0 +1,359 @@
+            while i < self.len {
+                // SAFETY: the element at `i` is valid.
+                unsafe {
+                    drop_fn(self.buf_at(i).as_ptr());
+                }
+                i += self.vtable.layout.size();
+            }
+        }
+        self.len = 0;
+    }
+
+    /// Clears the queue and frees the underlying buffer.
+    pub fn clear_and_shrink(&mut self) {
+        self.clear();
+        if self.cap > 0 && self.vtable.layout.size() != 0 {
+            // SAFETY: `buf` contains a valid unaliased allocation with the
+            // given size and alignment.
+            unsafe {
+                std::alloc::dealloc(
+                    self.buf.as_ptr(),
+                    Layout::from_size_align_unchecked(self.cap, self.vtable.layout.align()),
+                );
+            }
+            self.cap = 0;
+        }
+    }
+
+    fn grow(&mut self) {
+        assert!(self.vtable.layout.size() != 0, "zst overflow");
+        let align = self.vtable.layout.align();
+        let (new_cap, buf) = if self.cap == 0 {
+            let element_size = self.vtable.layout.size();
+            let new_cap = if element_size >= 256 {
+                element_size
+            } else {
+                element_size * 4
+            };
+            // SAFETY: `new_cap` is non-zero and at least as big as `align`,
+            // which is a power of 2.
+            let buf =
+                unsafe { std::alloc::alloc(Layout::from_size_align_unchecked(new_cap, align)) };
+            (new_cap, buf)
+        } else {
+            let new_cap = self.cap.checked_mul(2).unwrap();
+            // SAFETY: `buf` is a valid allocation with the given layout, and
+            // `new_cap` is non-zero.
+            let buf = unsafe {
+                std::alloc::realloc(
+                    self.buf.as_ptr(),
+                    Layout::from_size_align_unchecked(self.cap, align),
+                    new_cap,
+                )
+            };
+            (new_cap, buf)
+        };
+        let Some(buf) = NonNull::new(buf) else {
+            // SAFETY: these layout parameters were validated above.
+            let layout = unsafe { Layout::from_size_align_unchecked(new_cap, align) };
+            std::alloc::handle_alloc_error(layout);
+        };
+        // Move the trailing elements to the end of the new buffer.
+        if self.len > 0 && self.head + self.len > self.cap {
+            let n = self.cap - self.head;
+            let new_head = new_cap - n;
+            // SAFETY: `buf` is valid for reads and writes at the given offsets.
+            unsafe {
+                std::ptr::copy(buf.as_ptr().add(self.head), buf.as_ptr().add(new_head), n);
+            }
+            self.head = new_head;
+        }
+        self.buf = buf;
+        self.cap = new_cap;
+    }
+}
+
+impl Drop for ErasedVecDeque {
+    fn drop(&mut self) {
+        self.clear_and_shrink();
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
this might be a useful datastructure to fuzz. i.e: declare a few "interesting" types, and then have the fuzzer run a bunch of random operations with them